### ISSUE TITLE
openssl 1.1.0c raises compiler warning on DEFINE_LHASH_OF(OPENSSL_CSTRING);

### ISSUE THREAD

**Issue by kdekker** [openssl/openssl#2214]
OpenSSL 1.1.0c raises a compiler warning (that in our case was turned into an error) on:

openssl\lhash.h(198): error C4090: 'function': different 'const' qualifiers

This can be fixed by adding an const qualifier to OPENSSL_LH_insert() in the same header file. The diff is then:
```
75c75
< void *OPENSSL_LH_insert(OPENSSL_LHASH *lh, void *data);
---
> void *OPENSSL_LH_insert(OPENSSL_LHASH *lh, const void *data);
```
Adding this, will solve a lot of compiler warning. The cause for this warning is that OPENSSL_CSTRING is already const (char *).

Please let me know whether this information is enough to fix this issue. If needed, I can deliver a diff in patch format.

**Comment by richsalz** [comment]
That change isn't sufficient, since the declaration and the function definition won't agree and the file will fail to compile.  A more complete PR that added const more completely would be useful, but not clear if we'd do it for 1.1.1 as there isn't consistency on what causes an ABI breakage.

**Comment by kdekker** [comment]
You are right, my mistake.

However, now, both cURL as well a lot else raises on Windows:

w:\common\thirdparty\openssl\build\windows\x64\include\openssl\lhash.h(198): error C4090: 'function': different 'const' qualifiers

As we have turned C4090 into an error, it became an error instead of a warning. Usually, mixing const and non const is (in our code) considered as a bug. That's why I need to fix this, otherwise our code won't build anymore while using (part of) openSSL.

Questions (please forgive my ignorance):
Q1: What is a PR?
Q2: If you like, I can create another patch-style diff for the involved source file (I will then also check that everything of openSSL 1.1.0c compiles well, on a wide variety of platforms (Solaris. AIX, Linux, HPUX, Windows).

Regards,
Kees

________________________________
Van: Rich Salz [notifications@github.com]
Verzonden: woensdag 11 januari 2017 18:41
Aan: openssl/openssl
CC: Kees Dekker; Author
Onderwerp: Re: [openssl/openssl] openssl 1.1.0c raises compiler warning on DEFINE_LHASH_OF(OPENSSL_CSTRING); (#2214)


That change isn't sufficient, since the declaration and the function definition won't agree and the file will fail to compile. A more complete PR that added const more completely would be useful, but not clear if we'd do it for 1.1.1 as there isn't consistency on what causes an ABI breakage.

**Comment by richsalz** [comment]
We don't do cURL. :)  A PR is a pull request.  See the gihub overview docs.
It might be that the most useful way to address this is to turn off C4090 as an error for files that use the openssl stuff.

**Comment by kdekker** [comment]
I understand. But turning off, is turning off for a whole file, not only for OpenSSL. I just tried to get our developer to work conscientious :-)

Note that the warning is not raised on one line (lhash.h:197). The root cause is IMO that OPENSSL_STRING is non const (no warning) and OPENSSL_CSTRING is const.

I understand that you are not doing cURL, but I need to get all together working (including libssh2/zlib/curl/openSSL). So remaining comptible is very nice...

________________________________
Van: Rich Salz [notifications@github.com]
Verzonden: woensdag 11 januari 2017 21:42
Aan: openssl/openssl
CC: Kees Dekker; Author
Onderwerp: Re: [openssl/openssl] openssl 1.1.0c raises compiler warning on DEFINE_LHASH_OF(OPENSSL_CSTRING); (#2214)


We don't do cURL. :) A PR is a pull request. See the gihub overview docs.
It might be that the most useful way to address this is to turn off C4090 as an error for files that use the openssl stuff.

**Comment by kdekker** [comment]
I have checked the code in crypto\lhash\lhash.c and see that simply changing to const void * for the second argument is not possible, if a new element is inserted. The return type of then also must change. However, based on lhash.h, OPENSSL_LH_INSERT() is not called directly, but only indirectly (by using static inline functions) that always cast the return value. This code seems to ask for C++ overloading :).
And unfortunately, using OPENSSL_API_COMPAT < 0x10100000L maps lh_insert also to OPENSSL_LH_INSERT, causing changing return type a break in the API.

I think I have no other possibility to follow your advice. There is still something wrong in using a hash table with const pointers (like OPENSSL_CSTRING does) combined with this code. But I don't know. You may need to add a OPENSSL_LH_insert_const() function to address this + DEFINE_LHASH_CONST_OF() ? This function is then identical to OPENSSL_LH_insert() except of the const input and the const return type. There is already some stuff in lhash.h that deals with the const things.

I don't have enough knowledge about openSSL to propose the 'right' fix.

Regards,
Kees

From: Kees Dekker
Sent: Wednesday, January 11, 2017 22:04
To: openssl/openssl <reply@reply.github.com>; openssl/openssl <openssl@noreply.github.com>
Cc: Author <author@noreply.github.com>
Subject: RE: [openssl/openssl] openssl 1.1.0c raises compiler warning on DEFINE_LHASH_OF(OPENSSL_CSTRING); (#2214)

I understand. But turning off, is turning off for a whole file, not only for OpenSSL. I just tried to get our developer to work conscientious :-)

Note that the warning is not raised on one line (lhash.h:197). The root cause is IMO that OPENSSL_STRING is non const (no warning) and OPENSSL_CSTRING is const.

I understand that you are not doing cURL, but I need to get all together working (including libssh2/zlib/curl/openSSL). So remaining comptible is very nice...

________________________________
Van: Rich Salz [notifications@github.com]
Verzonden: woensdag 11 januari 2017 21:42
Aan: openssl/openssl
CC: Kees Dekker; Author
Onderwerp: Re: [openssl/openssl] openssl 1.1.0c raises compiler warning on DEFINE_LHASH_OF(OPENSSL_CSTRING); (#2214)

We don't do cURL. :) A PR is a pull request. See the gihub overview docs.
It might be that the most useful way to address this is to turn off C4090 as an error for files that use the openssl stuff.

-

**Comment by richsalz** [comment]
Yes it’s hard and painful.  Thanks for looking at it.

**Comment by kdekker** [comment]
Just curious: will anyone look for it to solve this? Or do I something else to mark this as bug?
If you like, I can make a reproduction. But the root cause (at our side) is that warning C4090 is turned into an error (with reason, for all our code). I can suppress this for OpenSSL, but still find this a bug...

**Comment by FdaSilvaYY** [comment]
See #1939

**Comment by mattcaswell** [comment]
Looks like this was fixed by #3420.
